pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -302,6 +305,84 @@ func (m *kubeGenericRuntimeManager) getKubeletContainerStatuses(podSandboxID str
 	return statuses, nil
 }
 
+// generateFailedContainerEvent generates a failed event for the container.
+func (m *kubeGenericRuntimeManager) generateFailedContainerEvent(containerID kubecontainer.ContainerID, podName, reason, message string) {
+	ref, ok := m.containerRefManager.GetRef(containerID)
+	if !ok {
+		glog.Warningf("No ref for pod %q", podName)
+		return
+	}
+	m.recorder.Event(ref, api.EventTypeWarning, reason, message)
+}
+
+// killContainer kills a container through the following steps:
+// * Run the pre-stop lifecycle hooks (if applicable).
+// * Stop the container.
+func (m *kubeGenericRuntimeManager) killContainer(pod *api.Pod, containerID kubecontainer.ContainerID, containerSpec *api.Container, reason string, gracePeriodOverride *int64) error {
```

Comment:
Can we pull out the pre-stop hook execution part into a function to improve the readability?